src/machine/machine_gameboyadvance.go

```diff
@@ -187,6 +220,157 @@ func (DisplayMode3) Display() error {
+func (SpriteRAM) Enable2D() {
+	IO.LCD.DISPCNT.ClearBits(DISPCNT_OBJ_VRAM_MAP_1D)
+	IO.LCD.DISPCNT.SetBits(DISPCNT_DISPLAY_OBJ)
+}
+
+// Enable1D sets the display to assume 1D sprites.
+func (SpriteRAM) Enable1D() {
+	IO.LCD.DISPCNT.SetBits(DISPCNT_DISPLAY_OBJ | DISPCNT_OBJ_VRAM_MAP_1D)
+}
+
+// Disable disables sprite rendering.
+func (SpriteRAM) Disable() {
+	IO.LCD.DISPCNT.ClearBits(DISPCNT_DISPLAY_OBJ)
+}
+
+// SpriteAttrs map the Object Attribute registers for a single sprite.
+type SpriteAttrs struct {
+	Y volatile.Register16 // Y offset, mode, flags, bitdepth, shape, etc
+	X volatile.Register16 // X offset, flipping, size, etc
+	T volatile.Register16 // Tile offset, priority, etc
+	_ volatile.Register16 // affine transform interleaved
+}
+
+// Sprite*Attrs constants help the compiler ensure that these otherwise-unitless
+// values are only passed for the correct configuration options.
+type (
+	SpriteYAttrs uint16
+	SpriteXAttrs uint16
+)
+
+// Sprite Attribute Constants
+const (
+	// Bit Depth (selected by SetupN, no need to pass them in)
+	SPRITE_4BPP = 0 << 13
+	SPRITE_8BPP = 1 << 13
+
+	// Offsets (add to tile index if using TILE_BLOCK_S1)
+	SPRITE_4BPP_S1_OFFSET = 512
+	SPRITE_8BPP_S1_OFFSET = 256
+
+	// Shape (Y Atribute)
+	SPRITE_SQUARE SpriteYAttrs = 0 << 14
+	SPRITE_WIDE   SpriteYAttrs = 1 << 14
+	SPRITE_TALL   SpriteYAttrs = 2 << 14
+
+	// Size (X Attribute)
+	SPRITE_SIZE_S  SpriteXAttrs = 0 << 14
+	SPRITE_SIZE_M  SpriteXAttrs = 1 << 14
+	SPRITE_SIZE_L  SpriteXAttrs = 2 << 14
+	SPRITE_SIZE_XL SpriteXAttrs = 3 << 14
+
+	// Flipping (X Attribute, non-affine only)
+	SPRITE_FLIP_H SpriteXAttrs = 1 << 12
+	SPRITE_FLIP_V SpriteXAttrs = 1 << 13
+)
+
+// Setup4 sets up a 4bpp sprite.
+//
+// Note that if your tiles are in the second sprite tile block, you will need to
+// add SPRITE_4BPP_S1_OFFSET before passing it in.
+func (s *SpriteAttrs) Setup4(priority, tile, bank int, y SpriteYAttrs, x SpriteXAttrs) {
+	s.Y.Set(uint16(y))
+	s.X.Set(uint16(x))
+	s.T.Set(uint16(bank&0xF)<<12 | uint16(priority&0x3)<<10 | uint16(tile&0x3FF))
+}
+
+// Setup4 sets up a 8bpp sprite.
+//
+// The tile index will be automatically scaled for 8bpp tiles, so this index
+// will match the one used for the tile block.
+//
+// Note that if your tiles are in the second sprite tile block, you will need to
+// add SPRITE_8BPP_S1_OFFSET before passing it in.
+func (s *SpriteAttrs) Setup8(priority, tile int, y SpriteYAttrs, x SpriteXAttrs) {
+	tile *= 2 // tile offset counts Tile4s
+	s.Y.Set(uint16(y | SPRITE_8BPP))
+	s.X.Set(uint16(x))
+	s.T.Set(uint16(priority&0x3)<<10 | uint16(tile&0x3FF))
+}
+
+// TODO(kevlar): helpers for setting setup fields after setup:
+//   - SetFlip
+
+// SetPos sets the position of the sprite.
+//
+// The x coordinate is in the range [0,512).
+// The y coordinate is in the range [0,256).
+func (s *SpriteAttrs) SetPos(x, y int) {
+	s.Y.ClearBits(0xFF)
+	s.Y.SetBits(uint16(y & 0xFF))
+	s.X.ClearBits(0xFF)
```

Comment:
This looks like a bug.  The `x` coordinate is 9 bits, but it's only clearing the first 8 bits.